scripts/idl/tests/outputs/cluster_struct_attribute/cpp/BridgeClustersImpl.h

```diff
@@ -0,0 +1,98 @@
+#include <app-common/zap-generated/cluster-id.h>
+#include <app-common/zap-generated/attribute-id.h>
+
+#include <new>
+
+namespace clusters {
+
+struct DemoClusterCluster : public CommonCluster
+{
+  struct ArmFailSafeRequest
+  {
+    PrimitiveType<uint16_t, 2, ZCL_INT16U_ATTRIBUTE_TYPE> expiryLengthSeconds;
+    PrimitiveType<uint64_t, 8, ZCL_INT64U_ATTRIBUTE_TYPE> breadcrumb;
+    PrimitiveType<uint32_t, 4, ZCL_INT32U_ATTRIBUTE_TYPE> timeoutMs;
+  };
+
+
+  static constexpr chip::ClusterId kClusterId = ZCL_DEMO_CLUSTER_CLUSTER_ID;
+
+  chip::ClusterId GetClusterId() override { return kClusterId; }
+
+  EmberAfStatus Read(const EmberAfAttributeMetadata * am, uint8_t * buffer, uint16_t maxReadLength) override
+  {
+    switch(am->attributeId) {
+    case 5:
+      return mSingleFailSafe.Read(am, buffer, maxReadLength);
+    case 100:
+      return mArmFailsafes.Read(am, buffer, maxReadLength);
+    default:
+      return EMBER_ZCL_STATUS_FAILURE;
+    }
+  }
+
+  EmberAfStatus Write(const EmberAfAttributeMetadata * am, uint8_t * buffer) override
+  {
+    switch(am->attributeId) {
+    case 5:
+      return mSingleFailSafe.WriteFromMatter(am, buffer, this);
+    case 100:
+      return mArmFailsafes.WriteFromMatter(am, buffer, this);
+    default:
+      return EMBER_ZCL_STATUS_FAILURE;
+    }
+  }
+
+  void WriteFromBridge(chip::AttributeId attributeId, const uint8_t * buffer) override
+  {
+    switch(attributeId) {
+    case 5:
+      mSingleFailSafe.WriteFromBridge(buffer, this);
+      break;
+    case 100:
+      mArmFailsafes.WriteFromBridge(buffer, this);
+      break;
+    }
+  }
+
+  template<typename T>
+  void AddAllAttributes(T *list)
+  {
+    list->Add(mSingleFailSafe);
+    list->Add(mArmFailsafes);
+  }
+
+  chip::Span<const EmberAfAttributeMetadata> GetAllAttributes() override
+  {
+    return chip::Span<const EmberAfAttributeMetadata>(kAllAttributes);
+  }
+
+
+  Attribute<5, ATTRIBUTE_MASK_WRITABLE, StructType<ArmFailSafeRequest>> mSingleFailSafe;
+  Attribute<100, ATTRIBUTE_MASK_WRITABLE, ArrayType<1, StructType<ArmFailSafeRequest>>> mArmFailsafes;
+
+  static constexpr const EmberAfAttributeMetadata kAllAttributes[] = {
+    { 5, ZCL_STRUCT_ATTRIBUTE_TYPE, sizeof(ArmFailSafeRequest), ATTRIBUTE_MASK_WRITABLE | ZAP_ATTRIBUTE_MASK(EXTERNAL_STORAGE), ZAP_EMPTY_DEFAULT() },
+    { 100, ZCL_ARRAY_ATTRIBUTE_TYPE, sizeof(ArmFailSafeRequest) * 1 + 2, ATTRIBUTE_MASK_WRITABLE | ZAP_ATTRIBUTE_MASK(EXTERNAL_STORAGE), ZAP_EMPTY_DEFAULT() },
+  };
+};
+
+struct ClusterInfo
+{
+  chip::ClusterId id;
+  const char *name;
+  uint16_t size;
+  CommonCluster* (*ctor)(void*);
+} const kKnownClusters[] = {
```

Comment:
So, these are concrete cluster impls that mix both the metadata about clusters, as well as their storage for attributes, stored in in a database-like array. Why is this mixing both state and metadata? If I project out where this is going, it's going to create a super large array of cluster impls that have a ton of space reserved for state, which is not used at all right? (Since you actually use the lambda-based ctor to actually construct one of these instances later).

Reply:
This is just cluster metadata - id, name, sizeof, ctor, a few words per cluster type stored in .rodata. To make a device you'd allocate enough space from somewhere (static array, heap, wherever) to hold all its clusters and then construct all the clusters in place. Yes presumably this will end up with one entry per cluster type